File: go/bubble_sort.go
package main

import "fmt"

func main() {
	list := []int{5, 1, 4, 6, 3, 2}
	fmt.Println(list)
	fmt.Println(bubbleSort(list))
}

func bubbleSort(list []int) []int {
	sortedList := make([]int, len(list))
	// Clone list
	copy(sortedList, list)

	// Sort list
	for range sortedList {
		for idx, el := range sortedList[1:] {
			if sortedList[idx] > el {
				sortedList[idx+1] = sortedList[idx]
				sortedList[idx] = el
			}
		}
	}

	return sortedList
}

File: rust/bubble_sort.rs
fn main() {
	let list = [5, 1, 4, 6, 3, 2].to_vec();

    eprintln!("Bubble Sort");
    eprintln!("Original {:?}", list);
    eprintln!("Sorted {:?}", bubble_sort(list.clone()));
}

fn bubble_sort<T: Ord + Copy>(mut list: Vec<T>) -> Vec<T> {
    for i in 0 .. list.len() {
        for j in i .. list.len() {
            if list[i] > list[j] {
                list.swap(i, j);
            }
        }
    }

    return list;
}
